test/unit/ngsiv2/bidirectionalityHttp-test.js

```diff
@@ -0,0 +1,264 @@
+                    // payload of the mock.
+                    if (!body.expires)
+                    {
+                        return false;
+                    }
+                    else if (moment(body.expires, 'YYYY-MM-DDTHH:mm:ss.SSSZ').isValid())
+                    {
+                        expectedBody.expires = moment().add(config.deviceRegistrationDuration);
+                        var expiresDiff = moment(expectedBody.expires).diff(body.expires, 'milliseconds');
+                        if (expiresDiff < 500) {
+                            delete expectedBody.expires;
+                            delete body.expires;
+
+                            return JSON.stringify(body) === JSON.stringify(expectedBody);
+                        }
+
+                        return false;
+                    }
+                    else {
+                        return false;
+                    }
+                })
+                .reply(201, null, {'Location': '/v2/subscriptions/51c0ac9ed714fb3b37d7d5a8'});
+
+            contextBrokerMock
+                .matchHeader('fiware-service', 'smartGondor')
+                .matchHeader('fiware-servicepath', '/gardens')
+                .post('/v1/updateContext', utils.readExampleFile(
+                    './test/contextRequests/createBidirectionalDevice.json'))
+                .reply(200, '{}');
+
+            iotagentUl.start(config, function(error) {
+                request(provisionOptions, function(error, response, body) {
+                    done();
+                });
+            });
+        });
+
+        it('should return a 200 OK', function(done) {
+            request(notificationOptions, function(error, response, body) {
+                should.not.exist(error);
+                response.statusCode.should.equal(200);
+                done();
+            });
+        });
+
+        it('should leave the data in the polling queue', function(done) {
+            request(notificationOptions, function(error, response, body) {
+                iotAgentLib.commandQueue('smartGondor', '/gardens', 'MQTT_2', function(error, list) {
+                    should.not.exist(error);
+
+                    list.commands.length.should.equal(3);
+                    done();
+                });
+            });
+        });
+
+        it('should send all the data from the notification in command syntax', function(done) {
+            request(notificationOptions, function(error, response, body) {
+                iotAgentLib.commandQueue('smartGondor', '/gardens', 'MQTT_2', function(error, list) {
+                    var latitudeFound = false,
+                        longitudeFound = false;
+
+                    for (var i = 0; i < list.commands.length; i++) {
+
+                        if (list.commands[i].name === 'location' &&
+                            list.commands[i].type === 'geo:point' &&
+                            list.commands[i].value === '0,0') {
```

Comment:
Is this value correct? Looking to bidirectionalNotification.json I'd say it should be "12.4, -9.6".

If I'm right, what is a bit surprissing the unit test suite doesn't raise the error. Just to be sure... are you running "grunt test" before pushing each new commit?

Reply:
Yes, all changes has been passed by grunt test and individualy

   128 passing (22s)
  3 pending
Done, without errors.

In pending : 
 HTTP Transport binding: polling commands
 When a device asks for the list of commands and there is more than one command
      - should retrieve the list sepparated by the "#" character
When the device sends the result for multiple pending commands
      - should make all the updates in the Context Broker
When a device sends measures and command responses mixed in a message
      - should send both the command updates and the measures to the Context Broker

Reply:
Latest commit here: 457a1d8d4f910b14a887f6903803956374a834a5

Reply:
I'll try to explain my point with more detail... I have my code at `457a1d8` to have a common version of the code to set a common understanding.

At this case https://github.com/telefonicaid/iotagent-ul/blob/457a1d8d4f910b14a887f6903803956374a834a5/test/unit/ngsiv2/bidirectionalityHttp-test.js#L58, we have:

* Provisioning (from https://github.com/telefonicaid/iotagent-ul/blob/457a1d8d4f910b14a887f6903803956374a834a5/test/deviceProvisioning/provisionCommandBidirectional.json)

```
{
  "devices": [
    {
      "device_id": "MQTT_2",
      "entity_name": "Second MQTT Device",
      "entity_type": "AnMQTTDevice",
      "timezone": "America/Santiago",
      "protocol": "HTTP_UL",
      "transport": "HTTP",
      "attributes": [
        {
          "name":"location",
          "type":"geo:point",
          "expression": "${latitude}, ${longitude}",
          "reverse": [
            {
              "object_id":"latitude",
              "type": "string",
              "expression": "${trim(substr(@location, indexOf(@location, \",\") + 1, length(@location)))}"
            },
            {
              "object_id":"longitude",
              "type": "string",
              "expression": "${trim(substr(@location, 0, indexOf(@location, \",\")))}"
            }
          ]
        }
      ],
      "lazy": [],
      "static_attributes": [],
      "commands": []
    }
  ]
}
```

* Noficiation received at IOTA from CB (https://github.com/telefonicaid/iotagent-ul/blob/457a1d8d4f910b14a887f6903803956374a834a5/test/unit/ngsiv2/subscriptionRequests/bidirectionalNotification.json)

```
{
  "subscriptionId": "51c0ac9ed714fb3b37d7d5a8",
  "data": [
    {
        "location": {
            "type": "geo:point",
            "value": "12.4, -9.6"
          },
        "type": "AnMQTTDevice",
        "id": "Second MQTT Device"
    }
  ]
}
```

and as result what we have in the commands queue is (https://github.com/telefonicaid/iotagent-ul/blob/457a1d8d4f910b14a887f6903803956374a834a5/test/unit/ngsiv2/bidirectionalityHttp-test.js#L144):

```
                        if (list.commands[i].name === 'location' &&
                            list.commands[i].type === 'geo:point' &&
                            list.commands[i].value === '0,0') {
                            latitudeFound = true;
                        }

                        if (list.commands[i].name === 'latitude' &&
                            list.commands[i].type === 'string' &&
                            list.commands[i].value === '-9.6') {
                            latitudeFound = true;
                        }

                        if (list.commands[i].name === 'longitude' &&
                            list.commands[i].type === 'string' &&
                            list.commands[i].value === '12.4') {
                            longitudeFound = true;
}
```

If I understand correctly by your last comment, "grunt test" is validating correctly the above statement.

However, as far as I understand, the first command should use the value of location as received in the notification, that is: "12.4, -9.6". However "0,0" is got.

How this can be explianed? Maybe is thre some wrong in my argumentation?

Reply:
@dcalvoalonso @fermenreq did you see my last comment on this thread? What is the explanation or the wrong point in my argumentation?

Reply:
@fgalan , your argument was correct:

```
                        if (list.commands[i].name === 'location' &&
                            list.commands[i].type === 'geo:point' &&
                            list.commands[i].value === '0,0') {
                            latitudeFound = true;
                        }
```

In this point latitudeFound is obviously false. Nevertheless, the test did not correctly assess this value and in fact it override it later:

```
                        if (list.commands[i].name === 'latitude' &&
                            list.commands[i].type === 'string' &&
                            list.commands[i].value === '-9.6') {
                            latitudeFound = true;
                        }
```

Reply:
Fixed in  105051b102a576ed1a5a48e900522048366e5d01